base_layer/core/tests/tests/header_sync.rs

```diff
@@ -0,0 +1,373 @@
+            if let HeaderSyncStatus::Lagging(val) = sync_result.header_sync_status {
+                assert_eq!(val.best_block_header.height(), 0);
+                assert_eq!(val.reorg_steps_back, 0);
+            } else {
+                panic!("Should be 'Lagging'");
+            }
+        },
+        _ => panic!("Expected StateEvent::HeadersSynchronized event"),
+    }
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn test_header_sync_with_fork_happy_path() {
+    // env_logger::init(); // Set `$env:RUST_LOG = "trace"`
+
+    // Create the network with Alice node and Bob node
+    let (mut alice_state_machine, alice_node, bob_node, initial_block, consensus_manager, key_manager) =
+        sync::create_network_with_alice_and_bob_nodes().await;
+
+    // Add 1 block to Bob's chain
+    let bob_blocks =
+        sync::create_and_add_some_blocks(&bob_node, &initial_block, 1, &consensus_manager, &key_manager, &[3]).await;
+    assert_eq!(bob_node.blockchain_db.get_height().unwrap(), 1);
+
+    // Bob adds another block
+    let bob_blocks =
+        sync::create_and_add_some_blocks(&bob_node, &bob_blocks[1], 1, &consensus_manager, &key_manager, &[3]).await;
+    assert_eq!(bob_node.blockchain_db.get_height().unwrap(), 2);
+
+    // Alice adds 3 (different) blocks, with POW on par with Bob's chain, but with greater height
+    let _alice_blocks =
+        sync::create_and_add_some_blocks(&alice_node, &initial_block, 3, &consensus_manager, &key_manager, &[
+            3, 2, 1,
+        ])
+        .await;
+    assert_eq!(alice_node.blockchain_db.get_height().unwrap(), 3);
+    assert_eq!(
+        alice_node
+            .blockchain_db
+            .get_chain_metadata()
+            .unwrap()
+            .accumulated_difficulty(),
+        bob_node
+            .blockchain_db
+            .get_chain_metadata()
+            .unwrap()
+            .accumulated_difficulty()
+    );
+
+    // Alice attempts header sync, but POW is on par
+    let mut header_sync = sync::sync_headers_initialize_with_ping_pong_data(&alice_node, &bob_node);
+    let event = sync::sync_headers_execute(&mut alice_state_machine, &mut header_sync).await;
+    match event.clone() {
+        StateEvent::Continue => {
+            // Good - Header sync not attempted, sync peer does not have better POW
+        },
+        _ => panic!("Expected StateEvent::Continue event"),
+    }
+    // All is good, Bob is not banned
+    assert!(!sync::wait_for_is_peer_banned(&alice_node, bob_node.node_identity.node_id(), 1).await);
+
+    // Bob adds more blocks and draws ahead of Alice
+    let _bob_blocks =
+        sync::create_and_add_some_blocks(&bob_node, &bob_blocks[1], 2, &consensus_manager, &key_manager, &[3; 2]).await;
+    assert_eq!(bob_node.blockchain_db.get_height().unwrap(), 4);
+
+    // Alice attempts header sync, on a higher chain with less POW, headers will be lagging with reorg steps
```

Comment:
wait, This is unclear
If Bob's chain has more blocks, but less pow, Alice should not be lagging, or try syncing?